flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/partitioner/ForwardPartitioner.java

```diff
@@ -43,4 +44,9 @@ public int selectChannel(SerializationDelegate<StreamRecord<T>> record) {
 	public String toString() {
 		return "FORWARD";
 	}
+
+	@Override
+	public ChannelStateRescaler getDownstreamChannelStateRescaler() {
+		return ChannelStateRescaler.FIRST_CHANNEL;
```

Comment:
Is `FIRST_CHANNEL` here (i.e. during the assignment) the same as in `selectChannel` (i.e. during sending the data)?
In other words, I think this will put all `InputChannel` state into a single subtask, while it should go to the same subtasks as before the rescaling.
WDYT?

Reply:
Really good catch. I was confused by the selection process of ForwardPartitioner always selecting 0. But what we want is to have a round robin distribution (so assign same as before in case of upscaling), redistribute on downscale to avoid data loss.
I used this find to rethink the nomenclature and actually everything that is related to `ChannelStateRescaler` is actually about subtask instance mapping, so I renamed to `SubtaskStateMapper`. Then it becomes immediately apparent that choosing the first task is plain wrong here.

Reply:
Yes, `ROUND_ROBIN` makes sense here and the new naming looks good!